substrate/primitives/runtime/src/proving_trie/base16.rs

```diff
@@ -0,0 +1,327 @@
+	/// proof that are not needed.
+	pub fn create_multi_proof(&self, keys: &[Key]) -> Result<Vec<u8>, DispatchError> {
+		sp_trie::generate_trie_proof::<LayoutV1<Hashing>, _, _, _>(
+			&self.db,
+			self.root,
+			&keys.into_iter().map(|k| k.encode()).collect::<Vec<Vec<u8>>>(),
+		)
+		.map_err(|err| TrieError::from(*err).into())
+		.map(|structured_proof| structured_proof.encode())
+	}
+}
+
+impl<Hashing, Key, Value> ProvingTrie<Hashing, Key, Value> for BasicProvingTrie<Hashing, Key, Value>
+where
+	Hashing: sp_core::Hasher,
+	Key: Encode,
+	Value: Encode + Decode,
+{
+	/// Create a new instance of a `ProvingTrie` using an iterator of key/value pairs.
+	fn generate_for<I>(items: I) -> Result<Self, DispatchError>
+	where
+		I: IntoIterator<Item = (Key, Value)>,
+	{
+		let mut db = MemoryDB::default();
+		let mut root = Default::default();
+
+		{
+			let mut trie = TrieDBMutBuilderV1::new(&mut db, &mut root).build();
+			for (key, value) in items.into_iter() {
+				key.using_encoded(|k| value.using_encoded(|v| trie.insert(k, v)))
+					.map_err(|_| "failed to insert into trie")?;
+			}
+		}
+
+		Ok(Self { db, root, _phantom: Default::default() })
+	}
+
+	/// Access the underlying trie root.
+	fn root(&self) -> &Hashing::Out {
+		&self.root
+	}
+
+	/// Query a value contained within the current trie. Returns `None` if the
+	/// nodes within the current `MemoryDB` are insufficient to query the item.
+	fn query(&self, key: &Key) -> Option<Value> {
+		let trie = TrieDBBuilder::new(&self.db, &self.root).build();
+		key.using_encoded(|s| trie.get(s))
+			.ok()?
+			.and_then(|raw| Value::decode(&mut &*raw).ok())
+	}
+
+	/// Create a compact merkle proof needed to prove a single key and its value are in the trie.
+	///
+	/// This function makes a proof with latest substrate trie format (`LayoutV1`), and is not
+	/// compatible with `LayoutV0`.
+	fn create_proof(&self, key: &Key) -> Result<Vec<u8>, DispatchError> {
+		sp_trie::generate_trie_proof::<LayoutV1<Hashing>, _, _, _>(
+			&self.db,
+			self.root,
+			&[key.encode()],
+		)
+		.map_err(|err| TrieError::from(*err).into())
+		.map(|structured_proof| structured_proof.encode())
+	}
+
+	/// Verify the existence of `key` and `value` in a given trie root and proof.
+	///
+	/// Proofs must be created with latest substrate trie format (`LayoutV1`).
+	fn verify_proof(
+		root: &Hashing::Out,
+		proof: &[u8],
+		key: &Key,
+		value: &Value,
+	) -> Result<(), DispatchError> {
+		verify_proof::<Hashing, Key, Value>(root, proof, key, value)
+	}
+}
+
+impl<Hashing, Key, Value> ProofToHashes for BasicProvingTrie<Hashing, Key, Value>
+where
+	Hashing: sp_core::Hasher,
+	Hashing::Out: MaxEncodedLen,
+{
+	// This base 16 trie does not directly expose the depth of the trie, so we can roughly calculate
+	// it assuming the data in the proof are hashes, and the number of hashes present will tell us
+	// the depth of the trie.
+	fn proof_to_hashes(proof: &[u8]) -> Result<u32, DispatchError> {
```

Comment:
So the method takes a proof and return the depth of this proof?
The proof contains touched encoded touched existing node (that themselves contain hashes), so counting the number of those node for a proof return the exact depth (since proof are Vec<Vec<u8>>, with inner being encoded node, you probably just have to decode the compact len at the first position), actually should check each node since some are just value (when size > 32), but it should be just a 1 depth error. This is only if there is a single value access in the proof.
For compact proof it is mostly the same.
If we want to be correct (don't assume the proof contains a single value), then we need to parse the sparse contain trie: for proof the method would need to calculate all hash and root hash in param: very costy. For compact proof the root is always the first and we can go down the trie without recalculating hashes (we know the structure of node in proof from node that are not include : so it is quite doable).